bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java

```diff
@@ -367,24 +367,25 @@ void computeRuns (GC gc) {
 	for (int i=0; i<allRuns.length - 1; i++) {
 		StyleItem run = allRuns[i];
 		OS.SelectObject(srcHdc, getItemFont(run));
-		shape(srcHdc, run);
+		shape(gc, srcHdc, run);
 	}
 	SCRIPT_LOGATTR logAttr = new SCRIPT_LOGATTR();
 	SCRIPT_PROPERTIES properties = new SCRIPT_PROPERTIES();
-	int lineWidth = indent, lineStart = 0, lineCount = 1;
+	int lineWidth = DPIUtil.autoScaleUp(getDevice(), indent, gc.data.shell), lineStart = 0, lineCount = 1;
 	for (int i=0; i<allRuns.length - 1; i++) {
 		StyleItem run = allRuns[i];
 		if (tabs != null && run.tab) {
 			int tabsLength = tabs.length, j;
 			for (j = 0; j < tabsLength; j++) {
-				if (tabs[j] > lineWidth) {
-					run.width = tabs[j] - lineWidth;
+				int tab = DPIUtil.autoScaleUp(getDevice(), tabs[j], gc.data.shell);
+				if (tab > lineWidth) {
+					run.width = tab - lineWidth;
```

Comment:
I wouldn't do the scaling until necessary, so e.g. here scaling up lineWidth in 374 and tabs[j] in 380 is only necessary when assigning it to run.width in line 382, so I would probably not scale it before and change that to 
run.width = DPIUtil.autoScaleUp(getDevice(), tab - lineWidth, gc.data.shell)
Without having looked at the whole code, it could be scaling it up from the beginnging (as it currently is) makes more sense, but make sure it really does

Reply:
here line width is already scaled up. In that case, I'll have to remove scale up lineWidth before this as well.